deepspeed/runtime/zero/stage2.py

```diff
@@ -1575,6 +1583,27 @@ def _has_inf_or_nan(x, j=None):
                 return True
             return False
 
+    @staticmethod
+    def _async_has_inf_or_nan(x, async_buff, j=None):
+        try:
+            # if x is half, the .float() incurs an additional deep copy, but it's necessary if
+            # Pytorch's .sum() creates a one-element tensor of the same type as x
+            # (which is true for some recent version of pytorch).
+            cpu_sum = float(async_buff.copy_(x.float().sum(), non_blocking=True))
```

Comment:
This is a really good catch. One question I have here is, can we guarantee that the value of cpu_sum is the value of async_buff after the copy_ and not before? Since copy_ is non-blocking, wouldn't this cause a race condition, where float(async_buff) might get executed before copy_ finishes, in which case it will just get the value that was already in async_buff?

Reply:
I have a general question here since I have recently encoutered `non_blocking=True` a few times. What is the prescribed method for synchronizing this copy operation?

Reply:
Hi @samyam, thanks for catching the bug. Yeah, I thought there is an implicit stream synchronization in this case but realized Pytorch doesn't enforce anything. So as you said, it is getting a stale data in this case. I can enforce a stream synchronization by using `torch.cuda.Stream.synchronize(<stream>)` and this `<stream>` should be passed from all above the call stack where it is defined from. Of course, in this case the situation pretty much becomes similar to the case where there was no communication overlap at all (only slightly better than the baseline). If this `local_overflow` does not need to be updated immediately (sorry, I don't know the exact flow of the entire code yet), I think it's better to keep the reduce result as a `gpu_sum` and read it really when it's needed.

Reply:
Or a simpler way would be to pull out `update_overflow_tracker_for_param_grad` from the `copy_grads_in_partition` function and to create another for loop between line# 970 & 971 to simple pre-run `update_overflow_tracker_for_param_grad`. In this case `update_overflow_tracker_for_param_grad` is not located in between the copy operations and it will not hinder the asynchronous behavior. If someone can verify if the `update_overflow_tracker_for_param_grad` call is dependent to the previous loop iteration it would be nice.

Reply:
@tjruwase can synchronize  either with cuda.synchronize() or doing a wait_stream on the cuda stream. @shhssdm We cannot remove the update_overflow_tracker_from_param_grad from where it is because that is the only place where the grads we are using in the function are ready to be consumed, and the next statement removes those grads. But, can do something like self.gpu_sum = self.gpu_sum + x.float().sum(). Then we can use self.gpu_sum to calculate self.local_overflow in the has_overflow directly. This should be an equivalent way of figuring out self.local_overflow, and it would remove the synchronous copy_ in the middle of the communication overlap.

Reply:
@samyam Makes sense. I changed the code to cumulatively check the overflow instead if overlap_comm is enabled.